### ISSUE TITLE
leader election fails after network degration

### ISSUE THREAD

**Issue by hotstepper13** [mesosphere/marathon#4000]
Hey guys,

in some environment we use a multi master setup with 3 nodes. Sometimes we have nightly network performance degration where the response times went up and we can see that connections are closed. While our Infrastructure provider is analyzing this issue, we see that marathon loses the awareness of its leader state. The Mesos Stack is running fine and recovered from the network degration. Marathon still needs a full restart on all nodes to com back to live.

The Logs didnt provide much info:

```
Jun 19 02:28:48 server2 marathon[8049]: [2016-06-19 02:28:48,528] INFO Unable to read additional data from server sessionid 0x15556a6325b0015, likely server has closed socket, closing socket connection and attempting reconnect (org.apache.zookeeper.ClientCnxn:ZookeeperClient-watcherProcessor-SendThread(servername3:2181))
Jun 19 02:28:49 server2 marathon[8049]: [2016-06-19 02:28:49,483] INFO Opening socket connection to server servername2/masterip:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn:ZookeeperClient-watcherProcessor-SendThread(servername2:2181))
Jun 19 02:28:49 server2 marathon[8049]: [2016-06-19 02:28:49,483] INFO Socket connection established to servername2/masterip:2181, initiating session (org.apache.zookeeper.ClientCnxn:ZookeeperClient-watcherProcessor-SendThread(servername2:2181))
Jun 19 02:28:49 server2 marathon[8049]: [2016-06-19 02:28:49,485] INFO Session establishment complete on server servername2/masterip:2181, sessionid = 0x15556a6325b0015, negotiated timeout = 10000 (org.apache.zookeeper.ClientCnxn:ZookeeperClient-watcherProcessor-SendThread(servername2:2181))
Jun 19 02:28:50 server2 marathon[8049]: [2016-06-19 02:28:50,144] INFO Do not proxy to myself. Waiting for consistent leadership state. Are we leader?: false, leader: Some(masterip:8090) (mesosphere.marathon.api.LeaderProxyFilter$:qtp1456906673-2590)
Jun 19 02:28:50 server2 marathon[8049]: [2016-06-19 02:28:50,145] INFO Waiting for consistent leadership state. Are we leader?: false, leader: Some(masterip:8090) (mesosphere.marathon.api.LeaderProxyFilter$:qtp1456906673-2590)
Jun 19 02:28:50 server2 marathon[8049]: [2016-06-19 02:28:50,150] INFO Do not proxy to myself. Waiting for consistent leadership state. Are we leader?: false, leader: Some(masterip:8090) (mesosphere.marathon.api.LeaderProxyFilter$:qtp1456906673-2599)
Jun 19 02:28:50 server2 marathon[8049]: [2016-06-19 02:28:50,151] INFO Waiting for consistent leadership state. Are we leader?: false, leader: Some(masterip:8090) (mesosphere.marathon.api.LeaderProxyFilter$:qtp1456906673-2599)
```

We are running
Mesos: 0.28.1-2.0.20.ubuntu140
Marathon: 1.1.1-1.0.472.ubuntu140

It looks like that this is some sort of split brain avoidance. However, because mesos itself seems to recover from the network degration without any issue, I think there might be some room for improvements within marathon.

Any ideas?

Best regards
Frank

**Comment by gkleiman** [comment]
This should hopefully be fixed by 2a1bbe4.

**Comment by meichstedt** [comment]
Note: This issue has been migrated to https://jira.mesosphere.com/browse/MARATHON-1709. For more information see https://groups.google.com/forum/#!topic/marathon-framework/khtvf-ifnp8.
